src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets

```diff
@@ -2,24 +2,23 @@
 
   <PropertyGroup Condition="'$(RuntimeIdentifier)' != '' and '$(NETCoreSdkPortableRuntimeIdentifier)' != ''">
     <!-- Define the name of the runtime specific compiler package to import -->
-    <_targetOSPkg>$(NETCoreSdkPortableRuntimeIdentifier.SubString(0, $(NETCoreSdkPortableRuntimeIdentifier.LastIndexOf('-'))))</_targetOSPkg>
-    <_indexOfHyphen>$(_targetOSPkg.IndexOf('-'))</_indexOfHyphen>
-    <_targetOS Condition="'$(_indexOfHyphen)' &gt; -1">$(_targetOSPkg.SubString(0, $(_indexOfHyphen)))</_targetOS>
-    <_targetOS Condition="'$(_indexOfHyphen)' == -1">$(_targetOSPkg)</_targetOS>
-    <_indexOfPeriod>$(_targetOS.IndexOf('.'))</_indexOfPeriod>
-    <_targetOS Condition="'$(_indexOfPeriod)' &gt; -1">$(_targetOS.SubString(0, $(_indexOfPeriod)))</_targetOS>
+    <_hostOS>$(NETCoreSdkPortableRuntimeIdentifier.SubString(0, $(NETCoreSdkPortableRuntimeIdentifier.IndexOf('-'))))</_hostOS>
+
+    <_useTPI Condition="'$(TargetPlatformIdentifier)' != '' and '$(TargetPlatformIdentifier)' != 'unix'">true</_useTPI>
+    <_targetOS Condition="'$(_useTPI)' == 'true'">$(TargetPlatformIdentifier.ToLowerInvariant)</_targetOS>
+    <_targetOS Condition="'$(_useTPI)' != 'true'">$(_hostOS)</_targetOS>
     <_targetOS Condition="$(_targetOS.StartsWith('win'))">win</_targetOS>
 
+    <!-- On non-Windows, determine _hostArchitecture from NETCoreSdkPortableRuntimeIdentifier -->
+    <_hostArchitecture Condition="'$(_targetOS)' != 'win'">$(NETCoreSdkPortableRuntimeIdentifier.SubString($([MSBuild]::Add($(NETCoreSdkPortableRuntimeIdentifier.LastIndexOf('-')), 1))))</_hostArchitecture>
+    <!-- On Windows, default to host that matches OS architecture on Windows. It allows compilation using msbuild.exe that is x86 by default -->
+    <_hostArchitecture Condition="'$(_targetOS)' == 'win'">$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture.ToString().ToLowerInvariant)</_hostArchitecture>
+
     <!-- Determine _targetArchitecture from RuntimeIdentifier -->
     <_targetArchitecture>$(RuntimeIdentifier.SubString($([MSBuild]::Add($(RuntimeIdentifier.LastIndexOf('-')), 1))))</_targetArchitecture>
 
-    <RuntimeIlcPackageName>runtime.$(_targetOS)-$(_targetArchitecture).Microsoft.DotNet.ILCompiler</RuntimeIlcPackageName>
-
-    <!-- Default to host that matches OS architecture on Windows. It allows compilation using msbuild.exe that is x86 by default -->
-    <IlcHostArch Condition="'$(OS)' == 'Windows_NT' and '$(IlcHostArch)' == ''">$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture.ToString().ToLowerInvariant)</IlcHostArch>
```

Comment:
I was convenient to be able to override `IlcHostArch`. It helps with testing for and working around host-architecture specific bugs.

Reply:
We can still override it with `/p:_hostArchitecture=..` for rare (testing) cases. The underscore values is just a convention for "internal use". They are not enforced by MSBuild, unless we explicitly reset the value with: https://github.com/dotnet/runtime/blob/64688a6e0f6e73440d0b42ce554cc0b25e003b5c/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L33

Reply:
> We can still override it with /p:_hostArchitecture=..

Do we need `_hostArchitecture != ''` in all conditions that set `_hostArchitecture` to make this work?

> The underscore values is just a convention for "internal use".

FWIW, we have been using `Ilc` prefix to tag undocumented properties in native AOT SDK and to avoid potential collisions with similar names in the rest of the SDK.

Reply:
When property is passed from command-line, then it is read-only unless we explicitly reset it. But if we set it in project file, then it will have no effect without `_hostArchitecture != ''` condition. If we only want to support this scenario for ad-hoc testing, then we can live with command-line only option.

> FWIW, we have been using `Ilc` prefix to tag undocumented properties in native AOT SDK and to avoid potential collisions with similar names in the rest of the SDK.

I will add back the prefix.

Reply:
> When property is passed from command-line, then it is read-only unless we explicitly reset it.

I did not know that. Thanks.